signbank/dictionary/csv_import.py

```diff
@@ -721,3 +721,157 @@ def confirm_import_qualtrics_csv(request):
+                group_gloss_count[row["group"]] = 1
+            glosses.append(row["idgloss"].split(":")[1])
+
+    except csv.Error as e:
+        # Can't open file, remove session variables
+        request.session.pop("group_row_map", None)
+        request.session.pop("glosses", None)
+        # Set a message to be shown so that the user knows what is going on.
+        messages.add_message(request, messages.ERROR, _("Cannot open the file:" + str(e)))
+        return render(request, "dictionary/import_manual_validation_csv.html",
+                      {"import_csv_form": CSVFileOnlyUpload()}, )
+    except UnicodeDecodeError as e:
+        # File is not UTF-8 encoded.
+        messages.add_message(request, messages.ERROR, _("File must be UTF-8 encoded!"))
+        return render(request, "dictionary/import_manual_validation_csv.html",
+                      {"import_csv_form": CSVFileOnlyUpload()}, )
+
+    # Store dataset's id and the list of glosses to be added in session.
+    request.session["group_row_map"] = group_row_map
+    request.session["glosses"] = list(set(glosses))
+
+    return render(request, "dictionary/import_manual_validation_csv_confirmation.html",
+                  {"group_row_map": group_row_map, "group_gloss_count": group_gloss_count})
+
+
+@login_required
+@permission_required("dictionary.import_csv")
+@transaction.atomic()
+def confirm_import_manual_validation(request):
+    """This view adds the data to database if the user confirms the action"""
+    if not request.method == "POST":
+        # If request method is not POST, redirect to the import form
+        return HttpResponseRedirect(reverse("dictionary:import_manual_validation_csv"))
+
+    if "cancel" in request.POST:
+        # If user cancels adding data, flush session variables
+        request.session.pop("group_row_map", None)
+        request.session.pop("glosses", None)
+        # Set a message to be shown so that the user knows what is going on.
+        messages.add_message(request, messages.WARNING, _("Cancelled adding CSV data."))
+        return HttpResponseRedirect(reverse("dictionary:import_manual_validation_csv"))
+
+    if not "confirm" in request.POST:
+        return HttpResponseRedirect(reverse("dictionary:import_manual_validation_csv"))
+
+    manual_validation_aggregations = []
+    missing_glosses = []
+
+    if "group_row_map" and "glosses" in request.session:
+        gloss_pk_list = request.session["glosses"]
+        gloss_dict = Gloss.objects.in_bulk(gloss_pk_list)
+
+        gloss_row_map = request.session["group_row_map"]
+
+        # Go through csv data
+        for group, rows in gloss_row_map.items():
+            for row in rows:
+                gloss = gloss_dict.get(int(row["idgloss"].split(":")[1]))
+                if not gloss:
+                    missing_glosses.append((group, row["idgloss"]))
+                    continue
+                sign_seen_yes = row["yes"]
```

Comment:
**suggestion**: we haven't validated the contents of the cells in each row, so if the user provides a value we can't assign to a positive integer, the server would raise a server error instead of a validation error.

n.b.: it would be better to raise an issue on the initial upload page than on the confirmation page.